compiler/frontend/src/org/jetbrains/kotlin/resolve/DescriptorResolver.java

```diff
@@ -1094,6 +1094,11 @@ scopeForDeclarationResolution, container, false, new TraceBasedLocalRedeclaratio
                     substitutedSuperType = approximatingSuperType;
                 }
 
+                if (languageVersionSettings.supportsFeature(LanguageFeature.KeepNullabilityWhenApproximatingLocalType)) {
```

Comment:
We probably need to preserve original type flexibility, too
But I don't think should be a deprecation for that part

Reply:
I'm not entirely sure it's possible to have a flexible type here, need to think about this some more.

Reply:
Yeah, it is.

```
// FILE: J.java
public class J {
    public static <T> T flexibleId(T x) { return x; }
}
// FILE: main.kt
interface I {
    fun foo()
}

val x = J.flexibleId(object : I {
    override fun foo() {}
})

fun main() {
    x?.foo()
    x.foo()
}
```

FIR doesn't even approximate the type of `x` in this case.

Reply:
done